### PR TITLE
job-ingest: assign jobspec defaults before scheduler feasibility check

### PR DESCRIPTION
**Pull request by garlick**
Problem: jobspec defaults are assigned after the scheduler feasibility check, which deprives that check of important information like the queue assignment as noted in #4494.

Add a "pipeline stage" to job-ingest that comes before the validator, where the feasibility check is performed.  This stage starts a work crew of `flux job-frobnicator` tasks (python) that can load plugins that can modify jobspec in arbitrary ways.  Currently a "default" plugin is provided.    Thanks to @grondo for knocking that part out!  This replaces the `jobspec-default` jobtap plugin, which runs too late.

The modified jobspec is written to the KVS so tools should be able to access stuff like the queue name without needing process `jobspec-update` events.

There's a substantial refactor of the job-ingest module here, which was needed to organize the work crews.  (I like how it turned out).

Still todo:

- [x] more testing
- [x] ingest configuration documentation
- [x] consider adding a  generic [policy] config check to job-manager since it's hard to handle reconfig errors from the frobnicator
- [ ] frobnicator plugin for setting queue constraints (though perhaps that could be another PR)
- [ ] consider whether this PR should be split up

### CODE REVIEW

**Comment by garlick** [comment]
Rebased on current master, squashed the fixups, and added some unit tests for some lower level code that is now split out of job-ingest.c.

**Comment by garlick** [comment]
To address the fact that the frobnicator is not in a position to validate the `[policy]` config (since it cannot prevent the instance from starting or return an error to `flux config reload`), and because several different components use `[policy]`, I added validation of that part of the config, which is defined in RFC 33, to the broker itself.

Rebased on current master and dropped the WIP.  This is probably ready for some review.

Note: if needed we could easily break out to separate PRs:
- fixes to conform to RFC 33
- broker config validation
- job-ingest refactor
- replacement of jobtap defaults plugin with frobnicator

**Comment by garlick** [comment]
Restarted inception builder due to a docker timeout.

**Review by chu11 — COMMENTED** [review]
just getting started, first review comments over the first 7 or 8 commits

**Comment by garlick** [comment]
Thanks @chu11 - I went ahead and addressed those comments, then squashed and rebased on current master (resolving a conflict).

**Comment by garlick** [comment]
Restarted bionic - gcc8,content-s3,distcheck builder after this (spurious?) failure
```
2022-09-13T14:16:50.3185873Z not ok 16 - flux-top can call itself recursively
2022-09-13T14:16:50.3186250Z FAIL: t2801-top-cmd.t 16 - flux-top can call itself recursively
2022-09-13T14:16:50.3186667Z #	
2022-09-13T14:16:50.3186850Z #		SHELL=/bin/sh &&
2022-09-13T14:16:50.3187129Z #		flux jobs &&
2022-09-13T14:16:50.3187465Z #		flux proxy $(cat jobid2) flux jobs -c1 -no {id} >expected.id &&
2022-09-13T14:16:50.3187783Z #		cat <<-EOF >recurse.in &&
2022-09-13T14:16:50.3187996Z #		{ "version": 2 }
2022-09-13T14:16:50.3188185Z #		[0.5, "i", "j"]
2022-09-13T14:16:50.3188358Z #		[1.0, "i", "j"]
2022-09-13T14:16:50.3188544Z #		[1.5, "i", "j"]
2022-09-13T14:16:50.3188724Z #		[2.0, "i", "k"]
2022-09-13T14:16:50.3188889Z #		[2.5, "i", "
2022-09-13T14:16:50.3189068Z #	"]
2022-09-13T14:16:50.3189243Z #		[3.25, "i", "q"]
2022-09-13T14:16:50.3189417Z #		[3.75, "i", "q"]
2022-09-13T14:16:50.3189595Z #		EOF
2022-09-13T14:16:50.3189907Z #		$runpty -o recurse.log --input=recurse.in flux top &&
2022-09-13T14:16:50.3190293Z #		grep -q $(echo $(cat expected.id) | sed "s/ƒ//") recurse.log
2022-09-13T14:16:50.3190532Z #	
```

File: src/modules/job-ingest/pipeline.c [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    else {
+        flux_future_t *f;
+        if (!(f = validate_job (pl, job, error)))
+            return -1;
+        *fp = f;
+    }
+    return 0;
+}
+
+int unpack_ingest_subtable (json_t *o,
```
Comment (chu11) [inline src/modules/job-ingest/pipeline.c:210]: static, local to file

**Review by chu11 — COMMENTED** [review]
comments through the `job-ingest: restart work crew on reconfig` commit.  Thought it was a good stopping point for now.

File: src/modules/job-ingest/test/job.c [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    rc = flux_msg_vpack (msg, fmt, ap);
+    va_end (ap);
+    if (rc < 0)
+        BAIL_OUT ("could not pack request");
+    if (flux_msg_set_cred (msg, cred) < 0)
+        BAIL_OUT ("could not set request cred");
+    return msg;
+}
+
+void test_job_flags_guest (void *sec, const char *J)
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/modules/job-ingest/test/job.c:50]: Good idea!

File: src/modules/job-ingest/job-ingest.c [hunk 3]
```diff
@@ -651,127 +576,30 @@ static void submit_cb (flux_t *h, flux_msg_handler_t *mh,
...
-            errprintf (&error, "Internal error: %s", e.text);
+            errprintf (&error, "Internal error: %s", json_error.text);
             errmsg = error.text;
             goto error;
         }
         if (!(f = workcrew_process_job (ctx->validate, o)))
             goto error;
-        if (flux_future_then (f, -1., validate_continuation, job) < 0)
+        if (flux_future_then (f, -1., validate_continuation, job) < 0
+            || flux_future_aux_set (f, "ctx", ctx, NULL) < 0)
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/modules/job-ingest/job-ingest.c]: That code ends up getting moved to `pipeline.c`.  In that file, `validate_job()` calls `job_json_object()`.

**Comment by garlick** [comment]
Thanks @chu11!  Good comments!   I fixed those issues and forced a push.  Also the I marked the first set of comments resolved.

**Review by chu11 — COMMENTED** [review]
these comments are through the new frobnicator command, nothing big.

As a general comment, should we document somewhere what the expected input to the frobnicator command should be.  like the `{"jobspec":..., "userid": ..., ...}` format?  I can't find where we do that for the job-validator, so perhaps we're punting b/c its a very internal to flux kinda thing?

(Note, haven't moved onto rest of commits yet, so its possible above is documented in later commit.  I just couldn't find equivalent for job validator and wondered if it was going to be missing for frobnicator too.)

File: src/bindings/python/flux/job/Jobspec.py [hunk 4]
```diff
@@ -371,6 +371,25 @@ def duration(self, duration):
...
+
+    @queue.setter
+    def queue(self, queue):
+        """
+        Set target submission queue
+        """
+        if not isinstance(queue, str):
+            raise TypeError("queue must be a string")
+        self.setattr("system.queue", queue)
+
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/bindings/python/flux/job/Jobspec.py:392]: I think that comment is correct, but I'll ask @grondo to comment as that was his description.

[the remark this answers is not in the document]
Comment (chu11) [reply src/bindings/python/flux/job/Jobspec.py:392]: ahhh I see what was intended now.

**Comment by garlick** [comment]
Addressed that batch of comments, squashed, and rebased on current master.  Thanks!

**Comment by garlick** [comment]
Weird, one bionic builder seems unhappy wtih loads of errors similar to this:
```
TypeError: initializer for ctype 'struct idset *' appears indeed to be 'struct idset *',
  but the types are different (check that you are not e.g. mixing up different ffi instances)
```
Other bionic builders are OK. I'm just going to restart that one since I have no idea what might be causing this.

**Comment by grondo** [comment]
I have seen something like that from time to time, only in CI. Not sure what might cause it, but a restart usually makes it go away. Perhaps some kind of rare problem caused by order of operations in the build?

File: t/t2241-policy-config.t [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	queues.x.policy.foo = 1
+	EOT
+	test_must_fail flux config reload
+'
+test_expect_success 'malformed queues.NAME.policy.limits.duration key fails' '
+	cat >config/policy.toml <<-EOT &&
+	queues.x.policy.limits.duration = 1
+	EOT
+	test_must_fail flux config reload
+'
```
Comment (chu11) [inline t/t2241-policy-config.t:100]: related to previous comment, need to cover allow-user/allow-guest?

**Review by chu11 — COMMENTED** [review]
Just a few more comments, but a few questions/comments.

Why default the frobnicator off by default?  It seems weird to me that it would be defaulted off and the validator defaulted on, even if both are configured with zero plugins and not technically doing anything?  It would seem to make more sense for them to be both enabled/disabled by default.
                                                                                                                                                             
A few commits seem like they should be bug fixes in a separate PR, such as:
- job-manager: require duration limit to be FSD                                                                                                              
- broker: validate RFC 33 config

**Comment by garlick** [comment]
> Why default the frobnicator off by default?

This was my thinking:

The validator is on by default because (at minimum) it ensures no invalid jobspec enters the system.

The frobnicator is off by default because it is only needed when [policy]  jobspec defaults are configured, and (coming soon) when queues, are configured.  Most likely only the system instance.  Hence, off by default to avoid the overhead, which is greater than the validator since it passes the modified jobspec back to `job-ingest`.

I think one could make an argument to disable the jobspec validator by default, since in a single user instance, any invalid jobspec that enters the system is the user's fault as well as their problem.  But that would be not be for this PR.

> A few commits seem like they should be bug fixes in a separate PR

Splitting off those two in particular would create some extra work since, assuming the new PR would come before this one, the`jobspec-default` jobtap plugin would have to be changed, as well as its tests, when it's only to be removed in this PR.

File: t/t2113-job-ingest-pipeline.t [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+'
+test_expect_success 'run a job to trigger work crew with new config' '
+	flux mini submit /bin/true
+'
+test_expect_success HAVE_JQ 'workers were restarted' '
+	flux module stats job-ingest >stats5.out &&
+	jq -r ".pipeline.frobnicator.pids[0]" <stats5.out >frob2.pid &&
+	jq -r ".pipeline.validator.pids[0]" <stats5.out >val2.pid &&
+	test_must_fail test_cmp frob.pid frob2.pid &&
+	test_must_fail test_cmp val.pid val2.pid
```
[the remark this answers is not in the document]
Comment (garlick) [reply t/t2113-job-ingest-pipeline.t:61]: Hmm, I'm not sure I agree.   To me, the quoting adds an ick factor.  Want to expand on why that's better?  I am ready to be convinced :-)

> use = to compare numbers, etc.

Did you mean  `-eq`?

File: src/broker/brokercfg.c [hunk 7]
```diff
@@ -34,6 +35,252 @@ struct brokercfg {
...
+                        "allow-user", &allow_user,
+                        "allow-group", &allow_group) < 0) {
+        errprintf (error, "%s: %s", key, jerror.text);
+        goto inval;
+    }
+    return  0;
+inval:
+    errno = EINVAL;
+    return -1;
+}
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/brokercfg.c:185]: I was being lazy since this isn't implemented yet but you are right.  Will fix.

**Comment by garlick** [comment]
Those changes were a little more involved than previous batches so I pushed them as fixups for review.  LMK if you'd like them squashed.

**Comment by garlick** [comment]
(oops just noticed I dropped a test in my fixup of `t2241-policy-config.t` - going to force push a change to that fixup in a sec.

File: t/t2113-job-ingest-pipeline.t (hunk 6 above)
[the remark this answers is not in the document]
Comment (chu11) [reply t/t2113-job-ingest-pipeline.t:61]: obviously just a style preference, but for me i think seeing the `!=` or `-gt` or whatever makes it more clear what the test is doing.  I think the `test_cmp` is fine, but doing a `test_must_fail` of `test_cmp` adds an extra thought to it.

> Did you mean -eq?

Oh yeah.

**Comment by chu11** [comment]
> Those changes were a little more involved than previous batches so I pushed them as fixups for review. LMK if you'd like them squashed.

Looks good, squash away

File: t/t2113-job-ingest-pipeline.t (hunk 6 above)
[the remark this answers is not in the document]
Comment (chu11) [reply t/t2113-job-ingest-pipeline.t:61]: oh wait, these are pids, so i guess can just do `-ne`, no need for the quoting.

[the remark this answers is not in the document]
Comment (garlick) [reply t/t2113-job-ingest-pipeline.t:61]: Also, an advantage of the `test_cmp` is that if the test fails you can see the actual and expected values in the test output.  Of course that could be arranged in any case with some diag output, but is fairly compact and simple.  I guess if you're not feeling too strongly about it I'd prefer to leave those alone.

**Comment by garlick** [comment]
> Looks good, squash away

Done.

**Review by chu11 — APPROVED** [review]
Whew, that was a long review!

**Comment by garlick** [comment]
Thank you!  Yes, this was a lot.  @grondo did you want to have a go or should I hold my breath and set merge-when-passing and we can sort out any problems later?

**Comment by grondo** [comment]
I think it will be good to get this merged so we can keep momentum on queues support. Sorry about the sloppiness in the frobnicator python code!

**Comment by garlick** [comment]
> I think it will be good to get this merged so we can keep momentum on queues support. Sorry about the sloppiness in the frobnicator python code!

OK thanks guys!  The frobnicator is frobtabulous, as was your review @chu11.  Setting MWP.

**Pull request merged**